client/js/src/cache_map.js

```diff
@@ -0,0 +1,312 @@
+		node.previous = this._head;
+		this._head = node;
+		return 1;
+	},
+	/**
+	* Description: Converts input params into tx_clocks, based on whether there exists
+	*			an associated list for the give key,value we either append the list or
+	*			create a brand new entry in the map. 
+	*	@public	
+	*/
+	put:function(read_time, value_time, table, key, value)
+	{
+		if(read_time.constructor == Number)
+			read_time = new tx_clock(read_time);
+		if(value_time.constructor == Number)
+			value_time = new tx_clock(value_time);
+
+		var k     = this._keyGen(key, table);
+		var array = this._map[k];
+		if(array === undefined)					//there exists no entry for that key
+			this._addList(read_time, value_time, table, key, value);
+		else
+			this._appendList(read_time, value_time, table, key, value);
+		
+	},
+
+	/**
+	*	Description: subroutine for creating new entry in map. Will generate
+	*			new object entry and insert into list, then place reference to 
+	*			list entry into a new list in the object map
+	*	@private
+	*/
+	_addList:function(read_time, value_time, table, key, value)
+	{
+		//create new entry in map
+		var k = this._keyGen(key, table);
+		
+		//create new entry in list
+		var list_entry = {
+			key:key, 
+			table:table, 
+			value_time:value_time, 
+			cached_time:read_time, 
+			value:value,
+			next:null,
+			previous:null
+		};
+
+		this._add(list_entry);		
+		this._map[k] = [this._head];		//insert reference to list object	
+		this._size++;
+		this._prune();
+		return;
+
+	},
+
+	/**
+	*	Description: subroutine for appending to an existing list for a specific key
+	*			Checks for an element in our list which contains a value_time equal
+	*			to the value_time we are inserting. If so, update the cached_time to 
+	*			the later of the element's cached_time and the read_time parameter.
+	*			If updating, promote list element to mru, otherwise create a new entry
+	*	@private
+	*/
+	_appendList:function(read_time, value_time, table, key, value)
+	{
+		var k = this._keyGen(key, table);
+		for(var i in this._map[k])			//for each key:table
+		{
+			if(+this._map[k][i].value_time == +value_time)
+			{
+				var optimal = new tx_clock(
+					Math.max(this._map[k][i].cached_time, read_time)
+					);
+				this._map[k][i].cached_time = optimal;
+				this._promote(this._map[k][i]);					//mru, so promote it
+				return 1;
+			}
+		}
+
+		var list_entry = {		
+			key:key, 
+			table:table, 
+			value_time:value_time, 
+			cached_time:read_time, 
+			value:value,
+			next:null,					//add will set next and prev pointers
+			previous:null
+		};
+		this._add(list_entry);
+		this._map[k].push(this._head);
+		this._size++;
+		this._prune();
+		return 1;
+
+		
+	},
+	
+	/**
+	*	Description: wrapper function to remove the lru object from list and map
+	*	@private
+	*/
+	_prune:function()
+	{
+		while(this._size > this._limit)
+			this._evictOne();		
+		
+	},
+
+	/**
+	*	Description: Designed to remove the lru element from the list and then remove the 
```

Comment:
We know this is the description, and we know that it was designed (giving you the benefit of the doubt). Jump straight to "Remove the...".